### DISCUSSION TITLE
Wallet rendering twice?

### DISCUSSION THREAD

**Discussion by ignaciolarranaga** [mercadopago/sdk-react#66] [Q&A] [answered]
Guys, some idea why this might be rendering twice?

```
export default function PayWithMercadoPago() {
 ...
 const handleSubmit = () => {
  return new Promise((resolve, reject) => {
    ...
    resolve(metadata.MERCADO_PAGO.preferenceId);
  });
 });

 return (
    <>
      {allowedPaymentMethods?.indexOf(PaymentMethod.MERCADO_PAGO) !== -1 && (
        <Wallet onSubmit={handleSubmit} locale={mpLocale} />
      )}
      {errorMessage && <ErrorMessage text={errorMessage} contactUs={true} />}
    </>
  );
}
```

<img width="768" alt="Screenshot 2023-07-24 at 13 33 31" src="https://github.com/mercadopago/sdk-react/assets/15932965/10f097e1-e206-4d42-b2c6-d81f943070c8">

<img width="256" alt="Screenshot 2023-07-24 at 13 33 40" src="https://github.com/mercadopago/sdk-react/assets/15932965/47dc61bd-de29-4592-a2e4-964cef99c4b3">

May be I missed to put some id on the wallet or a container?

**Comment by icaldana** [comment]
Hello @ignaciolarranaga!

It usually happens when the render method got called multiple times. Do you have special condition such as API calls inside useEffects, for example? One suggestion is to wait for those conditions to get satisfied before you allow the Wallet Button to render.

If possible, feel free to provide more details so we can help you debug this issue.

**Comment by ignaciolarranaga** [comment, marked as answer]
Thanks @icaldana. I found the problem.

I'm using NextJS so I can not really import the library directly (because of this: https://github.com/mercadopago/sdk-react/discussions/52). I tried the suggestion here: https://github.com/mercadopago/sdk-react/discussions/52#discussioncomment-6091598 but the **initMercadoPago loaded dynamically interferes with the wallet rendering triggering the re-render causing the buttons to duplicate**. I tried moving the initMercadoPago away from the wallet rendering but the ["dynamic approach"](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading#with-no-ssr) ultimately causes the same interference.

The best solution I was able to find was to [load the library dynamically](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading#with-external-libraries), then run the initialization and create the component.
Sharing in case useful:
```
export default function PayButtons({
  countryCode,
  copyId,
  inPersonPickup,
}: PayButtonsParams) {
  ...
  const [Wallet, setWallet] = useState<ReactNode>();

  useEffect(() => {
    ...
    import('@mercadopago/sdk-react').then(({ Wallet, initMercadoPago }) => {
      initMercadoPago(config.mercado_pago_public_key);
      setWallet(<Wallet onSubmit={handleSubmit} locale={mpLocale} />);
    });
  }, [locale]);

  return (
    <>
      {Wallet}
      ...
    </>
  );
}
```

**Reply by icaldana** [reply]
Hi @ignaciolarranaga, thank you for sharing! Yeah it is not the perfect solution but I am glad it works! By the way we have planned out tasks to clarify the behavior on NextJS framework.

Thank you so much!
